toonz/sources/tnztools/tool.cpp

```diff
@@ -794,6 +794,36 @@ QString TTool::updateEnabled() {
   TXshSimpleLevel *sl = xl ? xl->getSimpleLevel() : 0;
   int levelType       = sl ? sl->getType() : NO_XSHLEVEL;
 
+  // If not in Level editor, let's use our current cell from the xsheet to
+  // find the nearest level before it
+  if (levelType == NO_XSHLEVEL &&
+      !m_application->getCurrentFrame()->isEditingLevel()) {
+    int r0, r1;
+    xsh->getCellRange(columnIndex, r0, r1);
+    TXshCell cell = xsh->getCell(std::min(r1, rowIndex), columnIndex);
+    xl            = cell.isEmpty() ? 0 : (TXshLevel *)(&cell.m_level);
+    sl            = cell.isEmpty() ? 0 : cell.getSimpleLevel();
+    levelType     = cell.isEmpty() ? NO_XSHLEVEL : cell.m_level->getType();
+  }
```

Comment:
In general, this logic is for dealing with empty cells and ensuring we're always checking against the correct level type, which would be whatever level is just before the selected cell. Otherwise, on empty cells, it checks against the 1st level type which could be different.

For the most part, this will be used for a specific case that would arise if PR #2483 goes in. Currently the Type tool shows as valid for all levels and empty targets. On any empty cell the Type tool appears valid even for Raster levels despite the fact it does not work on that level type.

However, if PR #2483 goes in, the Type tool will become invalid for Raster levels.  If the 1st level is a Vector level followed by a Raster level, then on an empty cell after the Raster level, the Type tool would appear as valid  because the Vector version is tested by default and is valid on empty cells. This logic forces to check based on the Raster, not the Vector so the Type tool will now properly show as invalid.

Hope that makes sense.

Reply:
Hmm, for me it's still unclear regarding the tool's behavior with this PR in the following two points:

1. Under the situation below, what do you think will happen when you draw with the brush tool on the frame 5 ? And what will happen when you use the type tool on the frame 5 ?
![image](https://user-images.githubusercontent.com/17974955/51519579-bbedd680-1e64-11e9-92bf-5d8a19bab3b9.png)

For the brush tool, it seems to be depends on the `Autocreation` preference option. If the `Autocreation` is set to `Enabled` the created level depends on the `Default Level Type`. and if it is set to `Use Xsheet as Animation Sheet` , a new raster frame will be added to the level "Raster".
OTOH, for the type tool, regardless of the `Autocreation` option the tool always becomes invalid as if the tools is tried to be used to raster level. IMO they should be consistent between tools.

1. Furthermore, if you put another vector level at frame7, now the type tool becomes valid at frame 5 as if it is used to the vector level. 
![image](https://user-images.githubusercontent.com/17974955/51520897-497ef580-1e68-11e9-9135-f12771112969.png)

Reply:
Scenario 1: the brush will show a normal brush icon and will work normal regardless. This is because brush tool is configured for empty targets for all 3 level types.  (Side note: The tooloption's bar  will display settings for Vector instead of Raster until an image is actually created.  This should probably be fixed) 

As for the Type tool, **without** #2483, the Type tool will show the normal icon but when you try to use it, it will create a blank raster frame but won't allow you to enter text because it's not designed to work on Raster levels.  With #2483, the Type tool should show the invalid icon and not allow you to even create a raster frame when you try to use the tool.

I'll submit a change to modify this logic so it does **not** get applied when `Disabled` or `Enabled` preference option is used.  This really should only apply for `Use  Xsheet as Animation Sheet`.

Scenario 2:  I need to revise the logic.  It should be scanning backwards for the 1st non-empty cell to get the nearest level.   Currently if it's not in a cell after the last level, it's reverting back to testing the vector brush against an empty cell and showing as valid.